### DISCUSSION TITLE
How to ingest and parse Fortigate logs?

### DISCUSSION THREAD

**Discussion by 8inaryMata1eao** [Security-Onion-Solutions/securityonion#3429] [Unsupported Versions]
After a few weeks of reading the documentation for SO and Elasticsearch I am at the point where there is no where else to go.  We have logs coming from multiple fortigate firewalls that used to go to gray log and SO 16.  Graylog handled it perfectly and SO 16 I was able to utilized the Syslog then redirect to logstash with a pre-written parser.  Now with SO 2, the documentation states parsing is done with Elasticsearch, 

> Parsing
In Security Onion 2, Elasticsearch receives unparsed logs from Logstash or Filebeat. Elasticsearch then parses and stores those logs. Parsers are stored in /opt/so/conf/elasticsearch/ingest/. Custom ingest parsers can be placed in /opt/so/saltstack/local/salt/elasticsearch/files/ingest/. To make these changes take effect, restart Elasticsearch using so-elasticsearch-restart.

That said it seems my issue is a two part question.  Since SO 2 documentation doesn't seem to provide the instructions how to do so. It might be that some of these issues are assumed that the users are elasticsearch experts and knowledge is implied.  

The process, I started off by going to /opt/so/conf/elasticsearch/ingest/ and copying the syslog file.  I then worked on grok debugger to create two patterns which would parse the source and destination ip and port pairs.  Once the parser worked in the grok debugger on multiple logs I cleared the grok patters in the syslog file and placed the two patterns in the existing patters place, see below.

Existing "syslog"
`        "grok":
                {
                        "field": "message",
                        "patterns": [
                                        **"^<%{INT:syslog.priority}>%{DATA:syslog.timestamp} %{WORD:source.application}(\\[%{DATA:pid}\\])?: %{GREEDYDATA:real_message}$",
                                        "^%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.host} %{SYSLOGPROG:syslog.program}: CEF:0\\|%{DATA:vendor}\\|%{DATA:product}\\|%{GREEDYDATA:message2}$"**
                                    ],
`
NEW Fortigate Syslog.fort
`        "grok":
                {
                        "field": "message",
                        "patterns": [
                                        "^(<%{NUMBER:pri}>)?date=%{YEAR}-%{MONTHNUM}-%{MONTHDAY} time=%{TIME} devname=\"%{WORD:host}\" devid=\"%{WORD}\" logid=\"%{INT}\" type=\"%{WORD:type}\" subtype=\"%{DATA:subtype}\" (%{GREEDYDATA}) ?srcip=%{IP:source_ip} (%{GREEDYDATA}) ?srcport=%{NUMBER:source_port} (%{GREEDYDATA}) ?(dstip=%{IP:destination_ip}) ?(%{GREEDYDATA} )?dstport=%{NUMBER:destination_port} %{GREEDYDATA}$",
                                        "^(<%{NUMBER:pri}>)?date=%{YEAR}-%{MONTHNUM}-%{MONTHDAY} time=%{TIME} devname=\"%{WORD:host}\" devid=\"%{WORD}\" logid=\"%{INT}\" type=\"%{WORD:type}\" subtype=\"%{DATA:subtype}\" (%{GREEDYDATA}) ?(srcip=%{IP:source_ip}) ?(dstip=%{IP:destination_ip}) ?(%{GREEDYDATA}) ?srcport=%{NUMBER:source_port} (%{GREEDYDATA}) ?(%{GREEDYDATA} )?dstport=%{NUMBER:destination_port} %{GREEDYDATA}$"
                                    ],`

After that was completed I went and added this file to the location /`opt/so/saltstack/local/salt/elasticsearch/files/ingest/syslog.fort` and ran the required command `so-elasticsearch-restart`. Elasticsearch rebooted fine and the file was listed in the log output yet nothing was being parsed and indexed.  After that process I started looking at the discussion section and found `Syslogs from cisco ASA #2945: https://github.com/Security-Onion-Solutions/securityonion/discussions/2945` where user  J0hn-W1ick states he is trying to parse ASA logs and having no success.  Later in the thread user stevewillia and John talk about how Security Onion 16 worked perfectly for this and was much easier.  stevewillia recommends that J0hn adds the new parsing document to the logstash area of the global.sls. In summary, its is important to understand how from start to finish do we create parsers and add completely to the system to enable parsing of new logs.

1. What is the process of creating ingest parsers for eg fortigate?  
2. How do we activate the parsers in elasticsearch?

    Exact version"  2.3.30
    Install source. Security Onion ISO
    Install Distributed: 1 Manager, 1 SN, 8 Sensors
    Does so-status show all services running? Yes
    Do you get any failures when you run salt-call state.highstate from the node? No

**Comment by weslambert** [comment]
As you mentioned above, if you are modifying existing parsing, copy from the default to local `/opt/so/saltstack/local/salt/elasticsearch/files/ingest` directory and restart ES (on each node, using Salt, etc with `so-elasticsearch-restart`)

You can check to see if the pipeline you are trying to modify/add is loaded with:

`so-elasticsearch-pipelines-list`

You can check to see if events have passed through different plugins within the pipeline with:

`so-elasticsearch-pipline-stats $pipelinename`

Thanks,
Wes

**Reply by 8inaryMata1eao** [reply]
Wes thanks for getting back to me I appreciate the help.  
So I was curious about my modified "syslog" ingester.  Do have to change the "description" field? Also at the bottom I see it says "pipeline":       { "name": "common", do I need to change that or is it safe to leave it that way?

**Reply by weslambert** [reply]
The `description` field is really only for your reference -- the name of the pipeline will end up being the name of the file you create.  For example, a pipeline file with a name of `my.pipeline` will result in a pipeline of the name `my.pipeline` being loaded into Elasticsearch. The pipeline statements in the pipeline file help to route the event to pipelines deemed necessary for processing.  

For example, after traversing the initially defined pipeline, the event will proceed to enter the common pipeline for common cleanup tasks, like removing or renaming certain fields.  Other pipeline files may also have multiple pipeline statements to determine which pipelines an event might be routed to based on the evaluation of a conditional. 

What you could do in the case of syslog is either search the message using Filebeat, and tag it at the source, then direct it to the appropriate pipeline based off of that, or take the syslog pipeline like you have done, modify it to also look for the Fortigate patterns, or a partial match, then direct it to another dedicated pipeline for Fortigate logs.
